crates/noirc_evaluator/src/ssa_refactor/opt/constant_folding.rs

```diff
@@ -57,15 +60,47 @@ impl Context {
             self.push_instruction(function, block, instruction);
         }
 
-        let terminator =
-            function.dfg[block].unwrap_terminator().map_values(|value| self.get_value(value));
+        let terminator = function.dfg[block]
+            .unwrap_terminator()
+            .clone()
+            .map_values(|value| self.resolve_value(function, value));
 
         function.dfg.set_block_terminator(block, terminator);
         self.block_queue.extend(function.dfg[block].successors());
     }
 
-    fn get_value(&self, value: ValueId) -> ValueId {
-        self.values.get(&value).copied().unwrap_or(value)
+    /// Fetches the folded version of the given `ValueId`. If the value refers to an instruction
+    /// result, the folded value will have already been inserted into the cache. If it is an array
+    /// and it isn't already in the cache, it will be recursively resolved and added to the cache.
+    /// This is necessary because processing instruction results alone is not sufficient for
+    /// catching all arrays that have been affected by the pass.
+    fn resolve_value(&mut self, function: &mut Function, value: ValueId) -> ValueId {
```

Comment:
Can this function can be removed if we use `.set_value_from_id` on any updated results?

It seems either way that this fix would need to apply to every pass that changes ValueIds.

Reply:
Yes, true